src/palette.c

```diff
@@ -0,0 +1,987 @@
+        }
+        else
+        {
+            if (!gPaletteFade.delayCounter)
+            {
+                if (a1->ps_field_8 != a1->base->pst_field_A)
+                {
+                    u32 srcOffset = a1->srcIndex * a1->base->size;
+                    u8 i;
+
+                    for (i = 0; i < a1->base->size; ++i)
+                        gPlttBufferFaded[a1->baseDestOffset + i] = a1->base->src[srcOffset + i];
+                }
+            }
+        }
+    }
+}
+
+// not used
+static void sub_80709B4(struct PaletteStruct *a1)
+{
+    if (!a1->ps_field_9)
+    {
+        s32 val = a1->base->pst_field_B_5;
+
+        if (!val)
+        {
+            a1->srcIndex = 0;
+            a1->ps_field_8 = a1->base->pst_field_A;
+            a1->ps_field_9 = a1->base->pst_field_C;
+            a1->destOffset = a1->baseDestOffset;
+        }
+        else
+        {
+            if (val < 0)
+                return;
+            if (val > 2)
+                return;
+            ResetPaletteStructByUid(a1->base->uid);
+        }
+    }
+    else
+    {
+        --a1->ps_field_9;
+    }
+}
+
+void ResetPaletteStructByUid(u16 a1)
+{
+    u8 paletteNum = GetPaletteNumByUid(a1);
+    if (paletteNum != 16)
+        ResetPaletteStruct(paletteNum);
+}
+
+void ResetPaletteStruct(u8 paletteNum)
+{
+    sPaletteStructs[paletteNum].base = &gDummyPaletteStructTemplate;
+    sPaletteStructs[paletteNum].ps_field_4_0 = 0;
+    sPaletteStructs[paletteNum].baseDestOffset = 0;
+    sPaletteStructs[paletteNum].destOffset = 0;
+    sPaletteStructs[paletteNum].srcIndex = 0;
+    sPaletteStructs[paletteNum].ps_field_4_1 = 0;
+    sPaletteStructs[paletteNum].ps_field_8 = 0;
+    sPaletteStructs[paletteNum].ps_field_9 = 0;
+}
+
+void ResetPaletteFadeControl(void)
+{
+    gPaletteFade.multipurpose1 = 0;
+    gPaletteFade.multipurpose2 = 0;
+    gPaletteFade.delayCounter = 0;
+    gPaletteFade.y = 0;
+    gPaletteFade.targetY = 0;
+    gPaletteFade.blendColor = 0;
+    gPaletteFade.active = FALSE;
+    gPaletteFade.multipurpose2 = 0; // assign same value twice
+    gPaletteFade.yDec = FALSE;
+    gPaletteFade.bufferTransferDisabled = FALSE;
+    gPaletteFade.shouldResetBlendRegisters = FALSE;
+    gPaletteFade.hardwareFadeFinishing = FALSE;
+    gPaletteFade.softwareFadeFinishing = FALSE;
+    gPaletteFade.softwareFadeFinishingCounter = 0;
+    gPaletteFade.objPaletteToggle = 0;
+    gPaletteFade.deltaY = 2;
+}
+
+// not used
+static void sub_8070AFC(u16 uid)
+{
+    u8 paletteNum = GetPaletteNumByUid(uid);
+    if (paletteNum != 16)
+        sPaletteStructs[paletteNum].ps_field_4_1 = 1;
+}
+
+// not used
+static void sub_8070B28(u16 uid)
+{
+    u8 paletteNum = GetPaletteNumByUid(uid);
+    if (paletteNum != 16)
+        sPaletteStructs[paletteNum].ps_field_4_1 = 0;
+}
+
+// not used
+static u8 GetPaletteNumByUid(u16 uid)
+{
+    u8 i;
+
+    for (i = 0; i < 16; ++i)
+        if (sPaletteStructs[i].base->uid == uid)
+            return i;
+    return 16;
```

Comment:
Ascribe a define to this - either `NUM_PALETTE_STRUCTS` (and then use that in the declaration of `sPaletteStructs`) or `NELEMS(sPaletteStructs)`.

Reply:
Now  NUM_PALETTE_STRUCTS is there